### DISCUSSION TITLE
How to implement getCause() when implementing AbstractThrowableProblem?

### DISCUSSION THREAD

**Discussion by jehrenzweig-leagueapps** [micronaut-projects/micronaut-core#7259] [Q&A] [answered]
I'm just getting started with Micronaut (using Kotlin) & am trying to understand how to implement the `getCause()` function when creating a custom class that implements `AbstractThrowableProblem`:
```
class CustomProblemResponseBase : AbstractThrowableProblem(
	URI.create("about:blank"),
	"Unable to process request.",
	HttpStatusType(HttpStatus.UNPROCESSABLE_ENTITY),
	"An error occurred while attempting to process the HTTP request.",
) {
	@JsonIgnore
	override fun getCause(): Exceptional {
		TODO("Not yet implemented")
	}
}
```

I don't understand why I'm forced to implement `getCause()` and/or what I should be replacing `TODO("Not yet implemenented")`.  What if there isn't any underlying cause I want to describe -- what would the body of `getCause()` look like then?  It appears I have to return _something_ -- I can't return null.

It's entirely possible I'm missing something about how `AbstractThrowableProblem` is intended to be used.  Any advice is greatly appreciated -- thanks in advance!

**Comment by yawkat** [comment]
you don't have to override that method

**Reply by jehrenzweig-leagueapps** [reply]
If I comment out `override fun getCause(): Exceptional { ... }`, I immediately see the following error in IntelliJ:
```
Class 'CustomProblemResponseBase' is not abstract and does not implement abstract base class member 
public abstract fun getCause(): Exceptional! defined in org.zalando.problem.AbstractThrowableProblem
```

**Reply by yawkat** [reply]
Ah, this is a kotlin issue, see the discussion here https://github.com/zalando/problem/issues/145

**Reply by jehrenzweig-leagueapps** [reply]
That issue is closed, so it appears I need to override `getCause()` in my code, even though in my case I'm not interested in overriding the superclass method's behavior.  Not ideal, but also not a big deal.

However, I'm still curious if there are any examples around that show how one would override `getCause()` and what that code would look like.  I just want to understand a use case that would mandate writing override code.

**Reply by yawkat** [reply]
I don't think it makes sense to override getCause for this class (except for the kotlin bug that requires it)

**Reply by jehrenzweig-leagueapps** [reply]
Ok, good.  I'm still curious what type of class / use case would warrant overriding `getCause()`, and what that method would look like.  But at least I have an answer to my original question -- thank you.

**Comment by jehrenzweig-leagueapps** [comment, marked as answer]
TLDR: There's a Kotlin issue ( https://github.com/zalando/problem/issues/145 ) that requires `getCause()` be overridden when implementing `AbstractThrowableProblem` -- even if you want the default behaviour & don't have any need/desire to write your own `getCause()` implementation.  

Unfortunately, the Kotlin ticket was marked closed, so the workaround right now is to replace `TODO("Not yet implemented")` with `super.cause` in the body of the overridden `getCause()` method:
```
override fun getCause(): Exceptional? = super.cause
```
